jobs/jobs.go

```diff
@@ -4,71 +4,111 @@
 	}
+}
 
-	if dataRetentionInterface := ejobs.GetDataRetentionInterface(); dataRetentionInterface != nil {
-		jobs.DataRetention = dataRetentionInterface.MakeJob(s)
-	}
+func SetJobProgress(store store.Store, jobId string, progress int64) (bool, *model.AppError) {
+	var job *model.Job
 
-	return jobs
-}
+	if result := <-store.Job().Get(jobId); result.Err != nil {
+		return false, result.Err
+	} else {
+		job = result.Data.(*model.Job)
+	}
 
-func (jobs *Jobs) Start() *Jobs {
-	l4g.Info("Starting jobs")
+	job.Status = model.JOB_STATUS_IN_PROGRESS
+	job.Progress = progress
 
-	jobs.startOnce.Do(func() {
-		if jobs.DataRetention != nil && *utils.Cfg.DataRetentionSettings.Enable {
-			go jobs.DataRetention.Run()
+	if result := <-store.Job().UpdateOptimistically(job, model.JOB_STATUS_IN_PROGRESS); result.Err != nil {
+		return false, result.Err
+	} else {
+		if !result.Data.(bool) {
+			return false, nil
 		}
+	}
 
-		// go jobs.SearchIndexing.Run()
-	})
+	return true, nil
+}
 
-	jobs.listenerId = utils.AddConfigListener(jobs.handleConfigChange)
+func SetJobSuccess(store store.Store, job *model.Job) *model.AppError {
+	result := <-store.Job().UpdateStatus(job.Id, model.JOB_STATUS_SUCCESS)
+	return result.Err
+}
 
-	return jobs
+func SetJobError(store store.Store, job *model.Job) *model.AppError {
+	result := <-store.Job().UpdateStatus(job.Id, model.JOB_STATUS_ERROR)
+	return result.Err
 }
 
-func (jobs *Jobs) handleConfigChange(oldConfig *model.Config, newConfig *model.Config) {
-	if jobs.DataRetention != nil {
-		if !*oldConfig.DataRetentionSettings.Enable && *newConfig.DataRetentionSettings.Enable {
-			go jobs.DataRetention.Run()
-		} else if *oldConfig.DataRetentionSettings.Enable && !*newConfig.DataRetentionSettings.Enable {
-			jobs.DataRetention.Stop()
-		}
-	}
+func SetJobCanceled(store store.Store, job *model.Job) *model.AppError {
+	result := <-store.Job().UpdateStatus(job.Id, model.JOB_STATUS_CANCELED)
+	return result.Err
 }
 
-func (jobs *Jobs) Stop() *Jobs {
-	utils.RemoveConfigListener(jobs.listenerId)
+func RequestCancellation(store store.Store, job *model.Job) *model.AppError {
+	if result := <-store.Job().UpdateStatusOptimistically(job.Id, model.JOB_STATUS_PENDING, model.JOB_STATUS_CANCELED); result.Err != nil {
+		return result.Err
+	} else if result.Data.(bool) {
+		return nil
+	}
 
-	if jobs.DataRetention != nil && *utils.Cfg.DataRetentionSettings.Enable {
-		jobs.DataRetention.Stop()
+	if result := <-store.Job().UpdateStatusOptimistically(job.Id, model.JOB_STATUS_IN_PROGRESS, model.JOB_STATUS_CANCEL_REQUESTED); result.Err != nil {
+		return result.Err
+	} else if result.Data.(bool) {
+		return nil
 	}
-	// jobs.SearchIndexing.Stop()
 
-	l4g.Info("Stopped jobs")
+	return model.NewLocAppError("Jobs.RequestCancellation", "jobs.request_cancellation.status.error", nil, "id=" + job.Id)
+}
 
-	return jobs
+func CancellationWatcher(ctx context.Context, jobId string, store store.Store, cancelChan chan interface{}) {
+	running := true
```

Comment:
Could get rid of this variable entirely and just `return` to terminate the loop.